cpp/main.cc

```diff
@@ -14,31 +15,98 @@ string getStringCC() {
   return str.as<string>();
 }
 
-struct PointS {
-  float x;
-  float y;
+class Point {
+ public:
+  Point(float x, float y) : _x(x), _y(y) {}
+  Point() {}
+
+  float x() const { return _x; }
+  float y() const { return _y; }
+
+  void to_json(json& j) const { j = json{{"x", _x}, {"y", _y}}; }
+
+  void from_json(const json& j) {
+    j.at("x").get_to(_x);
+    j.at("y").get_to(_y);
+  }
+
+ private:
+  float _x = 0;
+  float _y = 0;
 };
 
-class Point : public PointS {
+void to_json(json& j, const Point& p) {
+  p.to_json(j);
+}
+
+void from_json(const json& j, Point& p) {
+  p.from_json(j);
+}
+
+enum class ActionType { Unknown, TouchStart, TouchMove };
+
+NLOHMANN_JSON_SERIALIZE_ENUM(ActionType,
+                             {
+                                 {ActionType::Unknown, "unknown"},
+                                 {ActionType::TouchStart, "touchstart"},
+                                 {ActionType::TouchMove, "touchmove"},
+                             });
+
+class Action {
  public:
-  Point(float x, float y) {
-    PointS::x = x;
-    PointS::y = y;
-  };
+  Action(ActionType type, Point point) : _type(type), _point(point) {}
+  Action(){};
+  ActionType type() const { return _type; };
+  const Point& point() const { return _point; }
+  void to_json(json& j) const { j = json{{"type", _type}, {"point", _point}}; }
```

Comment:
Can we use camelCase? I know google style is snake_case, but I can't really stand it.

When you break it up into files, can we use FullCaps.h case?

I like that searching over the codebase naturally matches both the filename and the exported class name and that we don't need to search with different casing for the JS and C++ sides of the codebase.